### PR TITLE
schedule, checker: add ruleChecker

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
Introduce `RuleChecker` to fix region according to placement rules.

### What is changed and how it works?
- add ruleChecker
- add tests
- adapt operator.Builder

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: server/schedule/checker/rule_checker.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+func (c *RuleChecker) fixRulePeer(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit) (*operator.Operator, error) {
+	// make up peers.
+	if len(rf.Peers) < rf.Rule.Count {
+		return c.addRulePeer(region, rf)
+	}
+	// fix down/offline peers.
+	for _, peer := range rf.Peers {
+		if c.isDownPeer(region, peer) || c.isOfflinePeer(region, peer) {
+			return c.replaceRulePeer(region, fit, rf, peer)
```
Comment (rleungx) [inline server/schedule/checker/rule_checker.go]: Do we need to distinguish these two types of operators?

File: server/schedule/checker/rule_checker.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	return false
+}
+
+func (c *RuleChecker) isOfflinePeer(region *core.RegionInfo, peer *metapb.Peer) bool {
+	store := c.cluster.GetStore(peer.GetStoreId())
+	if store == nil {
+		log.Warn("lost the store, maybe you are recovering the PD cluster", zap.Uint64("store-id", peer.StoreId))
+		return false
+	}
+	return !store.IsUp()
```
Comment (rleungx) [inline server/schedule/checker/rule_checker.go]: prefer to use `IsOffline`

File: server/schedule/checker/rule_checker.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+// Check checks if the region matches placement rules and returns Operator to
+// fix it.
+func (c *RuleChecker) Check(region *core.RegionInfo) *operator.Operator {
+	checkerCounter.WithLabelValues("rule_checker", "check").Inc()
+
+	fit := c.cluster.FitRegion(region)
+	if len(fit.RuleFits) == 0 {
+		// If the region matches no rules, the most possible reason is it spans across
+		// multiple rules.
+		return c.fixRange(region)
```
Comment (rleungx) [inline server/schedule/checker/rule_checker.go]: Do we need more metrics?

File: server/schedule/checker/rule_checker.go (hunk 2 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: Here we treat `tomestone` as offline too. (it happens when force bury a store before moing out all regions). `replicaChecker`'s behaviour is like this too: https://github.com/pingcap/pd/blob/4d65bbefbc6db6e92fee33caa97415274512757a/server/schedule/checker/replica_checker.go#L216-L218

File: server/schedule/checker/rule_checker.go (hunk 3 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: good idea.

File: server/schedule/checker/rule_checker_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		EndKeyHex:   "FF",
+		Role:        placement.Voter,
+		Count:       1,
+	})
+	s.cluster.AddLeaderRegionWithRange(1, "", "", 1, 2, 3)
+	op := s.rc.Check(s.cluster.GetRegion(1))
+	c.Assert(op, NotNil)
+	c.Assert(op.Len(), Equals, 1)
+	var split operator.SplitRegion
+	c.Assert(op.Step(0), FitsTypeOf, split)
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/schedule/checker/rule_checker_test.go]: fixed.

File: server/schedule/checker/rule_checker.go (hunk 1 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: fixed.

File: server/schedule/checker/rule_checker.go (hunk 3 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: added.

File: server/schedule/checker/rule_checker.go (hunk 2 above)
Comment (rleungx) [reply server/schedule/checker/rule_checker.go]: Yes, but this function named `isOfflinePeer` which may be a little bit confused.

Comment (disksing) [reply server/schedule/checker/rule_checker.go]: em. do you have better suggestion? (replicaChecker uses `checkOfflinePeer` as function name too...)

Comment (disksing) [reply server/schedule/checker/rule_checker.go]: I think IsOfflineOrTomestone looks too verbose 😢

File: server/schedule/checker/rule_checker.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+		return operator.CreatePromoteLearnerOperator("fix-peer-role", c.cluster, region, peer)
+	}
+	if region.GetLeader().GetId() == peer.GetId() && rf.Rule.Role == placement.Follower {
+		checkerCounter.WithLabelValues("rule_checker", "fix-leader-role").Inc()
+		for _, p := range region.GetPeers() {
+			if c.allowLeader(fit, p) {
+				return operator.CreateTransferLeaderOperator("fix-peer-role", c.cluster, region, peer.GetStoreId(), p.GetStoreId(), 0)
+			}
+		}
+		return nil, errors.New("no new leader")
```
Comment (nolouch) [inline server/schedule/checker/rule_checker.go]: same

File: server/schedule/checker/rule_checker.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+		return nil, errors.New("no store to add peer")
+	}
+	peer := &metapb.Peer{StoreId: store.GetID(), IsLearner: rf.Rule.Role == placement.Learner}
+	return operator.CreateAddPeerOperator("add-rule-peer", c.cluster, region, peer, operator.OpReplica)
+}
+
+func (c *RuleChecker) replaceRulePeer(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit, peer *metapb.Peer, status string) (*operator.Operator, error) {
+	store := SelectStoreToReplacePeerByRule(c.name, c.cluster, region, fit, rf, peer)
+	if store == nil {
+		return nil, errors.New("no store to add peer")
```
Comment (nolouch) [inline server/schedule/checker/rule_checker.go]: same

File: server/schedule/checker/rule_checker.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+			return op, nil
+		}
+	}
+	return c.fixBetterLocation(region, fit, rf)
+}
+
+func (c *RuleChecker) addRulePeer(region *core.RegionInfo, rf *placement.RuleFit) (*operator.Operator, error) {
+	checkerCounter.WithLabelValues("rule_checker", "add-rule-peer").Inc()
+	store := SelectStoreToAddPeerByRule(c.name, c.cluster, region, rf)
+	if store == nil {
```
Comment (nolouch) [inline server/schedule/checker/rule_checker.go]: How about adding a metrics about why nil operator created.

File: server/schedule/checker/rule_checker.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+	}
+	fs = append(fs, filters...)
+	store := selector.NewReplicaSelector(getRuleFitStores(cluster, rf), rf.Rule.LocationLabels).
+		SelectTarget(cluster, cluster.GetStores(), fs...)
+	return store
+}
+
+// SelectStoreToReplacePeerByRule selects a store to replace a region peer in order to fit the placement rule.
+func SelectStoreToReplacePeerByRule(scope string, cluster opt.Cluster, region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit, peer *metapb.Peer, filters ...filter.Filter) *core.StoreInfo {
+	rf2 := removePeerFromRuleFit(rf, peer)
```
Comment (nolouch) [inline server/schedule/checker/rule_checker.go]: What purpose for this,  don't want to remove `losematchPeer`?

File: server/schedule/checker/rule_checker.go (hunk 7 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: good idea.

File: server/schedule/checker/rule_checker.go (hunk 8 above)
Comment (disksing) [reply server/schedule/checker/rule_checker.go]: Similar to [replicaChecker.SelectBestReplacementStore](https://github.com/pingcap/pd/blob/9ae11e20321092edd436431b19124bcd4a9968a9/server/schedule/checker/replica_checker.go#L121-L125), `replace` is done by 
*copy region with the old peer removed + add a new peer to the copied region*

The difference is in `ruleChecker`, replace happens inside a rule (remove a peer, add another peer fits the same rule), so the replace process besomes
*copy RuleFit with the old peer removed + add a new peer to the copied RuleFit*

Comment (nolouch) [reply server/schedule/checker/rule_checker.go]: got

**Comment by disksing** [comment]
PTAL @rleungx

**Pull request merged**